### DISCUSSION TITLE
使用 VS Code 使用xelatex编译后，正文引用显示变量名而不是序号（已经按照文档配置）

### DISCUSSION THREAD

**Discussion by CNZWX** [BITNP/BIThesis#558] [Q&A] [answered]
使用 VS Code 使用xelatex编译后，正文引用显示变量名而不是序号。（以最新发布的本科生毕业论文为模版）
正文…… \cite{yuFeiJiZongTiDuoXueKeSheJiYouHuaDeXianZhuangYuFaZhanFangXiang2008}……\cite{Hajela2012Application}
编译出来后就只显示正文…… yuFeiJiZongTiDuoXueKeSheJiYouHuaDeXianZhuangYuFaZhanFangXiang2008……Hajela2012Application。
且\begin{bibprint}
\printbibliography[heading=none]
%这里报错LaTeX: Empty bibliography.LaTeX
\end{bibprint}

<img width="1440" alt="截屏2024-10-16 21 09 07" src="https://github.com/user-attachments/assets/24448531-d413-465d-8c1e-b37d9bd96760">
<img width="1440" alt="截屏2024-10-16 21 22 41" src="https://github.com/user-attachments/assets/c1cc4d7c-b1de-40aa-a671-47586cc9d736">
<img width="437" alt="截屏2024-10-16 21 23 07" src="https://github.com/user-attachments/assets/761fea39-cb29-4581-84f5-e12c20434106">

**Comment by YDX-2147483647** [comment]
请问具体是怎么配置 VSC 的？一种可行的做法请参考[在 VS Code 中使用 latexmk 编译 - 📃 编辑器配置与模板编译 | BIThesis](https://bithesis.bitnp.net/guide/configure-and-compile.html#%E5%9C%A8-vs-code-%E4%B8%AD%E4%BD%BF%E7%94%A8-latexmk-%E7%BC%96%E8%AF%91)。

**Comment by CNZWX** [comment]
您好！感谢深夜回复。配置VSC的代码是直接复制粘贴文档里给出的第二种“在VS Code中使用xelatex编译”的代码，也记得加了逗号。
如附件截图。
同时命令执行烂，选择的recipe方式是xelatex。
再次感谢！


***@***.***
***@***.***



&nbsp;




------------------&nbsp;原始邮件&nbsp;------------------
发件人:                                                                                                                        "BITNP/BIThesis"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2024年10月17日(星期四) 凌晨0:12
***@***.***&gt;;
***@***.******@***.******@***.***&gt;;
主题:&nbsp;Re: [BITNP/BIThesis] 使用 VS Code 使用xelatex编译后，正文引用显示变量名而不是序号（已经按照文档配置） (Discussion #558)





 
请问具体是怎么配置 VSC 的？一种可行的做法请参考使用 VS Code 撰写与编译 LaTeX 模板 - 📃 编辑器配置与模板编译 | BIThesis。

**Reply by CNZWX** [reply]
![Uploading 截屏2024-10-17 00.19.27.png…]()

**Reply by YDX-2147483647** [reply]
图似乎没上传上来，重发一下？

**Reply by YDX-2147483647** [reply]
> 选择的recipe方式是xelatex。

啊，我知道了，初次应该选“xelatex -> biber -> xelatex * 2”，这会走全流程。

参考文献有变化后，必须要用 biber 整理参考文献；没变化时可以只用 xelatex。

**Reply by CNZWX** [reply]
<img width="1440" alt="截屏2024-10-17 00 19 27" src="https://github.com/user-attachments/assets/d1ecef7d-78c1-4872-bc29-98bab0d32c66">

**Reply by CNZWX** [reply]
刚刚试了一下使用xelatex -> biber -> xelatex * 2。显示报错This is BibTeX, Version 0.99d (TeX Live 2024)
The top-level auxiliary file: main.aux
I found no \citation commands---while reading file main.aux
I found no \bibdata command---while reading file main.aux
I found no \bibstyle command---while reading file main.aux
(There were 3 error messages)
如截图
<img width="1440" alt="截屏2024-10-17 00 36 41" src="https://github.com/user-attachments/assets/320bc6fb-31a0-4b41-81a2-0b1798acf570">
另：我在根据文档配置环境的时候发现在macos终端输入biber --version时候显示：biber: Command Line Tools are not installed - run 'xcode-select --install' to install (errno=17)
【此MacTex为清华镜像下载的版本】
求解，感谢！

**Reply by YDX-2147483647** [reply]
好，那你那边的`settings.json`大概率没问题。

// 按<kbd>Alt</kbd>+<kbd>Shift</kbd>+<kbd>F</kbd>可以自动格式化来调整缩进。

**Reply by YDX-2147483647** [reply]
> 另：我在根据文档配置环境的时候发现在macos终端输入biber --version时候显示：biber: Command Line Tools are not installed - run 'xcode-select --install' to install (errno=17)

这是个有用的信息，但可能原因很多……如果你是刚刚新安装的，有可能重启电脑之后就好了。（不过我没用过 mac）

**Reply by YDX-2147483647** [reply]
VS Code 显示`main.tex`有问题，可以具体看看吗？或者白天再说也行。

另：可以考虑[清除缓存重新编译](https://bithesis-wiki.netlify.app/guide/commands.html#%E6%B8%85%E9%99%A4%E7%BC%93%E5%AD%98)，在终端运行`latexmk -c`。

**Reply by CNZWX** [reply]
刚刚重启了一下，`biber --version`仍旧无变化。我在考虑是否需要重装MacTex...
`main.tex`如截图，报错集中在第一行。源码没有动，是发布的模版“本科毕业论文”
深夜辛苦，如果白天您有时间的话，或许可以麻烦您回复，感谢！！
<img width="1440" alt="截屏2024-10-17 00 54 56" src="https://github.com/user-attachments/assets/bd4a3c60-23c1-497a-a3f4-90fd584652da">

**Reply by YDX-2147483647** [reply]
![图片](https://github.com/user-attachments/assets/910af093-9d00-4b00-a9c3-238f6eacbe94)

警告让重新运行 biber，你这次是用 xelatex -> biber -> xelatex * 2 吗？无论是不是，再运行一次呢？

---

另：插件显示的报错可能未及时更新；调试错误时直接在终端运行`latexmk`（绕过 VSC 手动编译）看输出，或者仍用 VSC 但检查“输出 → LaTeX Compiler”（就是你之前截图的地方）更准。

![图片](https://github.com/user-attachments/assets/bf9c4142-3c9c-4961-b1c2-d90b078b1cd2)

---

> 我在考虑是否需要重装MacTex...

卸载重装之前，可以先运行`tlmgr install biber`，看用 TeX Live manager 能否安装上。

<details>
<summary>人工智能的完整回答</summary>

如果你在MacTeX安装后使用`latexmk --version`命令正常，但`biber --version`提示“Command Line Tools are not installed”，这通常意味着biber的可执行文件没有被正确安装或者不在系统的PATH环境变量中。
以下是一些可能的解决方案：
1. **确认biber是否安装**：
   - 打开终端，尝试使用`which biber`命令来查看biber的安装路径。如果返回了路径，说明biber已经安装；如果没有返回路径，则可能需要重新安装。
2. **重新安装MacTeX**：
   - 有时候可能因为安装过程中的问题导致biber没有正确安装。你可以尝试重新安装MacTeX，确保在安装过程中选择了安装所有组件。
3. **手动安装biber**：
   - 如果重新安装MacTeX不起作用，你可以尝试手动安装biber。可以从CTAN（Comprehensive TeX Archive Network）下载biber的MacOS版本，然后将其放在一个目录中，并将该目录添加到PATH环境变量。
4. **更新命令行工具**：
   - 由于提示信息是“Command Line Tools are not installed”，你可以尝试通过运行`xcode-select --install`来安装或更新MacOS的命令行工具。
5. **检查PATH环境变量**：
   - 确保biber的安装路径被添加到PATH环境变量中。你可以通过运行`echo $PATH`来查看当前的PATH设置，并确保其中包含了biber的路径。
6. **使用MacTeX提供的包管理器**：
   - MacTeX通常包含了`tlmgr`（TeX Live Manager），你可以使用它来安装或更新biber。在终端中运行`tlmgr install biber`。
7. **重启计算机**：
   - 在安装或更新软件后，有时候需要重启计算机才能使更改生效。
如果以上步骤都没有解决问题，你可以尝试在终端中运行以下命令来查找biber的具体安装位置：
```bash
find /usr/local/texlive/ -name biber
```
如果找到了biber的可执行文件，你可以将其路径添加到PATH环境变量中。如果仍然找不到，你可能需要检查你的MacTeX安装是否完整或者是否需要重新安装。

</details>

**Reply by CNZWX** [reply]
后续：我下载了xtools，依旧不行。
但是
我换了硕博的模版后，点进去latexmk编译，一点问题都没有，什么改动都不需要。
【不确定是不是因为本科毕业论文的模版存在问题】
<img width="1440" alt="截屏2024-10-17 03 53 31" src="https://github.com/user-attachments/assets/40988a09-74b2-4444-832f-007c641a1ecf">

**Reply by YDX-2147483647** [reply]
我刚又试了一下，至少我这里 undergraduate-thesis 能正常编译；其实每次发布版本前，都会人工与上一版本生成的PDF比对，应该不太容易一打开就出错……
![图片](https://github.com/user-attachments/assets/f9b6468c-a199-42ba-82f9-9fa780326826)

能否另下载 undergraduate-thesis，编译一次后，把整个文件夹打包发过来呢？我仍然怀疑是缓存问题，但没有日志很难确定。

**Reply by CNZWX** [reply, marked as answer]
听了您的建议，重新下载了undergraduate-thesis。这一次latexmk编译直接通过了，报错3个，但是没有引用出现变量名的问题了。
报错与注释一致，是字体的问题。main.log如附件
[main.log](https://github.com/user-attachments/files/17404677/main.log)

**Reply by CNZWX** [reply]
而后重启VSC后，字体问题也消失了。crazy

**Reply by YDX-2147483647** [reply]
> 字体的问题

正常就会有那几个字体问题。模板给了个字体测试表格，有些版本（比如加粗斜体）字体本身没提供，LaTeX就会提示它换成别的了，具体和平台有关。实际使用时一般用不到那么多种字体。

**Reply by YDX-2147483647** [reply]
> 重启VSC后，字体问题也消失了

这恐怕是因为没有真正发生编译。插件会调用`latexmk`，从输出里筛出问题；重启 VSC 后，插件又调用`latexmk`，但`latexmk`发现文件没变，不用重新编译，所以插件就“忘了”那些问题。如果你随便改几个字，应该就又有了。

// 如果开发时要测试模板，我一般用`latexmk -gg`，它会先清除缓存再编译，相当于先运行上面说的`latexmk -c`再`latexmk`。

https://github.com/BITNP/BIThesis/blob/eb0adc8b0077b60d1d6f816b4d7a8231ab64112b/Makefile#L132

**Reply by CNZWX** [reply]
学到了，非常感谢您的帮助！！
